### ISSUE TITLE
ruler: Ruler Stuck on some queries with latency > interval.

### ISSUE THREAD

**Issue by bwplotka** [thanos-io/thanos#2397]
Detected by our new alert: 

```
alert: ThanosRuleTSDBNotIngestingSamples
  annotations:
    message: Thanos Rule {{$labels.job}} did not ingest any samples for the last 15
      minutes.
  expr: |
    sum by (job) (rate(prometheus_tsdb_head_samples_appended_total{job=~"thanos-rule.*"}[5m])) <= 0
  for: 10m
  labels:
    severity: critical
```
https://github.com/thanos-io/thanos/blob/master/examples/alerts/alerts.md

Investigating.

**Comment by bwplotka** [comment]
Any update on this @metalmatze ?

**Comment by bwplotka** [comment]
Right now we think it might be rather issue with DNS Discovery + no timeouts/keep lives + abrupted TCP connection.

**Comment by bwplotka** [comment]
Stuck CPU profile:
![image](https://user-images.githubusercontent.com/6950331/79481967-1514cd80-8008-11ea-8340-410f0ca1a618.png)

**Comment by bwplotka** [comment]
![image](https://user-images.githubusercontent.com/6950331/79489327-0253c600-8013-11ea-92d1-f522d26854ec.png)

**Comment by bwplotka** [comment]
This was unrelated to latency, rather to stuck gRPC on DNS updates. This was fixed on master from 0.13+
